cairo/ethereum/cancun/vm/interpreter.cairo

```diff
@@ -331,3 +358,212 @@ func _execute_code{
+        new MessageCallOutputStruct(
+            gas_left=Uint(0),
+            refund_counter=U256(new U256Struct(0, 0)),
+            logs=empty_tuple_log,
+            accounts_to_delete=empty_set1,
+            touched_accounts=empty_set2,
+            error=error,
+        ),
+    );
+    return msg;
+}
+
+// @dev Drop an `Evm` struct by squashing all of its fields except for Environment.
+func squash_evm{range_check_ptr, evm: Evm}() {
+    alloc_locals;
+    let evm_struct = cast(evm.value, EvmStruct*);
+
+    // Squash stack
+    let stack = evm.value.stack;
+    let stack_start = stack.value.dict_ptr_start;
+    let stack_end = cast(stack.value.dict_ptr, DictAccess*);
+    dict_squash(cast(stack_start, DictAccess*), cast(stack_end, DictAccess*));
+
+    // Squash memory
+    let memory = evm.value.memory;
+    let memory_start = memory.value.dict_ptr_start;
+    let memory_end = cast(memory.value.dict_ptr, DictAccess*);
+    dict_squash(cast(memory_start, DictAccess*), cast(memory_end, DictAccess*));
+
+    // No squash for Env which is an implicit argument in `process_message_call`
+
+    // Squash valid_jump_destinations
+    let valid_jump_destinations = evm.value.valid_jump_destinations;
+    let valid_jump_destinations_start = valid_jump_destinations.value.dict_ptr_start;
+    let valid_jump_destinations_end = cast(valid_jump_destinations.value.dict_ptr, DictAccess*);
+    dict_squash(cast(valid_jump_destinations_start, DictAccess*), valid_jump_destinations_end);
+
+    // Squash accounts_to_delete
+    let accounts_to_delete = evm.value.accounts_to_delete;
+    let accounts_to_delete_start = accounts_to_delete.value.dict_ptr_start;
+    let accounts_to_delete_end = cast(accounts_to_delete.value.dict_ptr, DictAccess*);
+    dict_squash(cast(accounts_to_delete_start, DictAccess*), accounts_to_delete_end);
+
+    // Squash touched_accounts
+    let touched_accounts = evm.value.touched_accounts;
+    let touched_accounts_start = touched_accounts.value.dict_ptr_start;
+    let touched_accounts_end = cast(touched_accounts.value.dict_ptr, DictAccess*);
+    dict_squash(cast(touched_accounts_start, DictAccess*), cast(touched_accounts_end, DictAccess*));
+
+    // Squash accessed_addresses
+    let accessed_addresses = evm.value.accessed_addresses;
+    let accessed_addresses_start = accessed_addresses.value.dict_ptr_start;
+    let accessed_addresses_end = cast(accessed_addresses.value.dict_ptr, DictAccess*);
+    dict_squash(cast(accessed_addresses_start, DictAccess*), accessed_addresses_end);
+
+    // Squash accessed_storage_keys
+    let accessed_storage_keys = evm.value.accessed_storage_keys;
+    let accessed_storage_keys_start = accessed_storage_keys.value.dict_ptr_start;
+    let accessed_storage_keys_end = cast(accessed_storage_keys.value.dict_ptr, DictAccess*);
+    dict_squash(
+        cast(accessed_storage_keys_start, DictAccess*), cast(accessed_storage_keys_end, DictAccess*)
+    );
+
+    return ();
```

Comment:
you need to re-instanciated an Evm instance, with all data structures being the squashed dicts (if you intend to re-use them, as you do in 

```

    let squashed_evm = evm;
    tempvar msg = MessageCallOutput(
        new MessageCallOutputStruct(
            gas_left=squashed_evm.value.gas_left,
            refund_counter=U256(new U256Struct(squashed_evm.value.refund_counter, 0)),
            logs=squashed_evm.value.logs,
            accounts_to_delete=squashed_evm.value.accounts_to_delete,
            touched_accounts=squashed_evm.value.touched_accounts,
            error=squashed_evm.value.error,
        ),
    );
```

otherwise, MessageCallOutput will not contain the squashed dicts

Reply:
another possibility is to simply make this function return a MessageCallOutput. Which could be easier.

Reply:
good catch 
Done [6bd1f17](https://github.com/kkrt-labs/keth/pull/539/commits/6bd1f17210704899da834c806808eefb1fa281a0)